### ISSUE TITLE
"not" operator precedence changed?

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#1932]
Migrated from [rt.perl.org#3202](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3202) (status was 'resolved')

Searchable as RT3202$

**Comment by p5pRT** [comment]
### From tmcd@jump.net

I had a script that worked fine\, or so it appeared\, under previous
Perl 5s\.  When I upgraded to 5\.006\, I got errors for the line

&nbsp;    if \(not \($id\, $count\) = /^\(\[^\,\]\+\)\,\\s\*\(\[0\-9\]\+\)$/ \) \{

It reported

&nbsp;    Useless use of private variable in void context at \./hst\.pl line 1557\.
&nbsp;    Can't modify not in scalar assignment at \./hst\.pl line 1557\,
&nbsp;    near "/^\(\[^\,\]\+\)\,\\s\*\(\[0\-9\]\+\)$/ \) "

When I changed it to

&nbsp;    if \(\! \(\($id\, $count\) = /^\(\[^\,\]\+\)\,\\s\*\(\[0\-9\]\+\)$/\) \) \{

it compiled fine\.  \(That does do what I want&#8203;: if it matches\, assign
the text matched in the parens to $id and $count\, and if not\, it
should do the "if" block\, right?\)

Did I screw up the precedence originally myself?  Is this a
misfeature?

Summary of my perl5 \(revision 5\.0 version 6 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.0\.36\, archname=i586\-linux
&nbsp;    uname='linux tmcd\.austin\.tx\.us 2\.0\.36 \#1 tue dec 29 13&#8203;:11&#8203;:13 est 1998 i586 unknown '
&nbsp;    config\_args='\-de'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=undef d\_sfio=undef uselargefiles=define
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, optimize='\-O2'\, gccversion=2\.7\.2\.3
&nbsp;    cppflags=''
&nbsp;    ccflags =' \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'
&nbsp;    stdchar='char'\, d\_stdstdio=define\, usevfork=false
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=4
&nbsp;    alignbytes=4\, usemymalloc=n\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lndbm \-lgdbm \-ldb \-ldl \-lm \-lc \-lposix \-lcrypt
&nbsp;    libc=/lib/libc\-2\.0\.7\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'


Characteristics of this binary \(from libperl\)&#8203;:
&nbsp;  Compile\-time options&#8203;: USE\_LARGE\_FILES
&nbsp;  Built under linux
&nbsp;  Compiled at Apr 26 2000 12&#8203;:26&#8203;:51
&nbsp;  %ENV&#8203;:
&nbsp;    PERL5LIB="/home/tmcd/lib/perl5"
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /home/tmcd/lib/perl5
&nbsp;    /usr/local/lib/perl5/5\.6\.0/i586\-linux
&nbsp;    /usr/local/lib/perl5/5\.6\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.6\.0/i586\-linux
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.6\.0
&nbsp;    /usr/local/lib/perl5/site\_perl
&nbsp;    \.

**Comment by p5pRT** [comment]
### From @tamias

On Wed\, May 03\, 2000 at 01&#8203;:51&#8203;:47PM \-0500\, Timothy A\. McDaniel wrote&#8203;:
> I had a script that worked fine\, or so it appeared\, under previous
> Perl 5s\.  When I upgraded to 5\.006\, I got errors for the line
> 
>     if \(not \($id\, $count\) = /^\(\[^\,\]\+\)\,\\s\*\(\[0\-9\]\+\)$/ \) \{
> 
> It reported
> 
>     Useless use of private variable in void context at \./hst\.pl line 1557\.
>     Can't modify not in scalar assignment at \./hst\.pl line 1557\,
>     near "/^\(\[^\,\]\+\)\,\\s\*\(\[0\-9\]\+\)$/ \) "
> 
> When I changed it to
> 
>     if \(\! \(\($id\, $count\) = /^\(\[^\,\]\+\)\,\\s\*\(\[0\-9\]\+\)$/\) \) \{
> 
> it compiled fine\.  \(That does do what I want&#8203;: if it matches\, assign
> the text matched in the parens to $id and $count\, and if not\, it
> should do the "if" block\, right?\)
> 
> Did I screw up the precedence originally myself?  Is this a
> misfeature?
> 

I believe this was an intentional change\.  It's documented in perldelta&#8203;:




---
=item Parenthesized not\(\) behaves like a list operator

The C\<not> operator now falls under the "if it looks like a function\,
it behaves like a function" rule\.

As a result\, the parenthesized form can be used with C\<grep> and C\<map>\.
The following construct used to be a syntax error before\, but it works
as expected now&#8203;:

&nbsp;    grep not\($\_\)\, @&#8203;things;

On the other hand\, using C\<not> with a literal list slice may not
work\.  The following previously allowed construct&#8203;:

&nbsp;    print not \(1\,2\,3\)\[0\];

needs to be written with additional parentheses now&#8203;:

&nbsp;    print not\(\(1\,2\,3\)\[0\]\);

The behavior remains unaffected when C\<not> is not followed by parentheses\.



---
Basically\, we've traded one instance of DWIM for another\.

Ronald
